gapic/src/main/com/google/gapid/perfetto/models/PowerSummaryTrack.java

```diff
@@ -0,0 +1,184 @@
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/** {@link Track} summarizing the total Power usage. */
+public class PowerSummaryTrack extends Track.WithQueryEngine<PowerSummaryTrack.Data> {
+  private static String viewSql;
+  private static int numPowerRailTracks;
+  public static Unit unit;
+  public static Double minValue;
+  public static Double maxValue;
+
+  private static final String VIEW_SQL_MONOTONIC =
+      "select ts + 1 ts, lead(ts) over win - ts dur, lead(value) over win - value value, lead(id)"
+          + " over win id from counter where track_id = ";
+
+  public PowerSummaryTrack(QueryEngine qe, int numTracks) {
+    super(qe, "power_sum");
+    this.numPowerRailTracks = numTracks;
+  }
+
+  public int getNumPowerRailTracks() {
+    return numPowerRailTracks;
+  }
+
+  public static Perfetto.Data.Builder enumerate(Perfetto.Data.Builder data) {
+    ImmutableListMultimap<String, CounterInfo> counters = data.getCounters(CounterInfo.Type.Global);
+    List<CounterInfo> powerRails =
+        counters.entries().stream()
+            .filter(entry -> entry.getKey().startsWith("power.rails"))
+            .map(Map.Entry::getValue)
+            .collect(Collectors.toList());
+
+    if (powerRails.size() == 0) {
+      return data;
+    }
+
+    // Power Group
+    viewSql = buildViewSql(powerRails);
+    PowerSummaryTrack powerSummaryTrack = new PowerSummaryTrack(data.qe, powerRails.size());
+
+    // Power Rails tracks.
+    for (CounterInfo powerRail : powerRails) {
+      unit = powerRail.unit;
+      CounterTrack powerRailTrack = new CounterTrack(data.qe, powerRail);
+      data.tracks.addTrack(
+          powerSummaryTrack.getId(),
+          powerRailTrack.getId(),
+          powerRail.name,
+          single(
+              state -> new CounterPanel(state, powerRailTrack, POWER_RAIL_COUNTER_TRACK_HEIGHT),
+              true,
+              true));
+    }
+
+    Group.UiFactory ui = group(state -> new PowerSummaryPanel(state, powerSummaryTrack), false);
+    data.tracks.addLabelGroup(null, powerSummaryTrack.getId(), "Power Usage", ui);
+    return data;
+  }
+
+  private static String buildViewSql(List<CounterInfo> powerRails) {
+    minValue = powerRails.get(0).min;
+    maxValue = powerRails.get(0).max;
+    StringBuilder sb =
+        new StringBuilder()
+            .append(
+                "select aggregateTable.ts ts, sum(aggregateTable.value) totalValue,"
+                    + " aggregateTable.dur dur from")
+            .append(" (")
+            .append(VIEW_SQL_MONOTONIC)
+            .append(powerRails.get(0).id)
+            .append(" window win as (order by ts) ")
+            .append(") as aggregateTable");
+    for (int i = 1; i < numPowerRailTracks; i++) {
+      sb.append("left join (")
+          .append(VIEW_SQL_MONOTONIC)
+          .append(powerRails.get(i).id)
+          .append(" window win as (order by ts) ")
```

Comment:
If you use the `String.format` pattern as used elsewhere in the Perfetto UI, you can have the entire query in the constant.

Reply:
Done, thanks for the catch!